### PR TITLE
First pass at Sequence API

### PR DESCRIPTION
**Pull request by ebolyen**
~~@jairideout is still combing through the last of the docstrings, so there may be some errors.~~ I'm done combing :)

Not looking for an extremely detailed review, we can go over this next week in person.

### CODE REVIEW

File: skbio/__init__.py [hunk 1]
```diff
@@ -13,18 +13,16 @@
...
-from skbio.tree import (
-    TreeNode, nj)
+from skbio.tree import TreeNode, nj
 from skbio.io import read, write
 
-__all__ = ['BiologicalSequence', 'NucleotideSequence', 'DNA', 'DNASequence',
-           'RNA', 'RNASequence', 'Protein', 'ProteinSequence',
+
+__all__ = ['Sequence', 'DNA',
+           'RNA', 'Protein',
```
Comment (ebolyen) [inline skbio/__init__.py]: Put on same line

File: skbio/alignment/_alignment.py [hunk 2]
```diff
@@ -585,73 +569,18 @@ def is_empty(self):
...
     def iteritems(self):
         """Generator of id, sequence tuples
 
         Returns
         -------
         generator of tuples
-            Each tuple contains ordered
-            (`skbio.sequence.BiologicalSequence.id`,
-            `skbio.sequence.BiologicalSequence`) pairs.
+            Each tuple contains ordered (`skbio.id`, `skbio.Sequence`) pairs.
```
Comment (ebolyen) [inline skbio/alignment/_alignment.py]: `skbio.id` -> `skbio.Sequence.id`

File: skbio/alignment/_ssw_wrapper.pyx [hunk 3]
```diff
@@ -791,13 +791,14 @@ def local_pairwise_align_ssw(sequence1, sequence2,
...
-            ProteinSequence(alignment.aligned_query_sequence, id='query'),
-            ProteinSequence(alignment.aligned_target_sequence, id='target')
+            Protein(alignment.aligned_query_sequence, id='query'),
+            Protein(alignment.aligned_target_sequence, id='target')
         ]
     else:
         seqs = [
-            NucleotideSequence(alignment.aligned_query_sequence, id='query'),
-            NucleotideSequence(alignment.aligned_target_sequence, id='target')
+            # TODO FIX THIS!
```
Comment (ebolyen) [inline skbio/alignment/_ssw_wrapper.pyx]: We should fix this.

File: skbio/io/fasta.py [hunk 4]
```diff
@@ -554,8 +550,8 @@
                             _format_fasta_like_records, _line_generator,
                             _too_many_blanks)
 from skbio.alignment import SequenceCollection, Alignment
-from skbio.sequence import (BiologicalSequence, NucleotideSequence,
-                            DNASequence, RNASequence, ProteinSequence)
+from skbio.sequence import (Sequence,
+                            DNA, RNA, Protein)
```
Comment (ebolyen) [inline skbio/io/fasta.py]: Same line

File: skbio/io/fastq.py [hunk 5]
```diff
@@ -273,8 +273,8 @@
                             _format_fasta_like_records, _line_generator,
                             _too_many_blanks)
 from skbio.alignment import SequenceCollection, Alignment
-from skbio.sequence import (BiologicalSequence, NucleotideSequence,
-                            DNASequence, RNASequence, ProteinSequence)
+from skbio.sequence import (Sequence,
+                            DNA, RNA, Protein)
```
Comment (ebolyen) [inline skbio/io/fastq.py]: same line

File: skbio/io/qseq.py [hunk 6]
```diff
@@ -120,12 +118,13 @@
...
 from future.builtins import zip, range
+from functools import partial
 
 from skbio.io import register_reader, register_sniffer, QSeqFormatError
 from skbio.io._base import _decode_qual_to_phred, _get_nth_sequence
 from skbio.alignment import SequenceCollection
-from skbio.sequence import (BiologicalSequence, NucleotideSequence,
-                            DNASequence, RNASequence, ProteinSequence)
+from skbio.sequence import (Sequence,
+                            DNA, RNA, Protein)
```
Comment (ebolyen) [inline skbio/io/qseq.py]: same line

File: skbio/io/tests/test_clustal.py [hunk 7]
```diff
@@ -227,6 +227,8 @@ def test_valid_alignment_to_clustal_and_clustal_to_alignment(self):
             testfile.close()
             testfile = open(fname, 'r')
             result_after = _clustal_to_alignment(testfile)
+            print(result_before)
+            print(result_after)
```
Comment (ebolyen) [inline skbio/io/tests/test_clustal.py]: remove print statements

File: skbio/io/tests/test_fasta.py [hunk 8]
```diff
@@ -11,17 +11,16 @@
...
+from skbio import (Sequence, DNA, RNA, Protein, SequenceCollection, Alignment)
 from skbio.io import FASTAFormatError
 from skbio.io.fasta import (
     _fasta_sniffer, _fasta_to_generator, _fasta_to_biological_sequence,
-    _fasta_to_nucleotide_sequence, _fasta_to_dna_sequence,
+    _fasta_to_dna_sequence,
     _fasta_to_rna_sequence, _fasta_to_protein_sequence,
     _fasta_to_sequence_collection, _fasta_to_alignment, _generator_to_fasta,
-    _biological_sequence_to_fasta, _nucleotide_sequence_to_fasta,
+    _biological_sequence_to_fasta,
```
Comment (ebolyen) [inline skbio/io/tests/test_fasta.py]: clean up block

File: skbio/io/tests/test_fasta.py [hunk 9]
```diff
@@ -395,7 +391,7 @@ def test_fasta_to_generator_valid_files(self):
         for exp, kwargs, fasta_fps, qual_fps in test_cases:
             for fasta_fp in fasta_fps:
                 obs = list(_fasta_to_generator(fasta_fp, **kwargs))
-
+                print(fasta_fp)
```
Comment (ebolyen) [inline skbio/io/tests/test_fasta.py]: remove print

File: skbio/io/tests/test_fastq.py [hunk 10]
```diff
@@ -12,9 +12,10 @@
 
 import unittest
 import warnings
+from functools import partial
 
-from skbio import (read, write, BiologicalSequence, NucleotideSequence,
-                   DNASequence, RNASequence, ProteinSequence,
+from skbio import (read, write, Sequence,
+                   DNA, RNA, Protein,
```
Comment (ebolyen) [inline skbio/io/tests/test_fastq.py]: clean up block

File: skbio/sequence/tests/test_nucleotide_sequence.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+        self.assertFalse(seq2.is_reverse_complement(seq1))
+
+    def test_is_reverse_complement_type_mismatch(self):
+        seq1 = ExampleNucleotideSequence('ABC')
+        seq2 = ExampleNucleotideSequenceSubclass('ABC')
+
+        with self.assertRaises(TypeError):
+            seq1.is_reverse_complement(seq2)
+
+# class NucelotideSequenceTests(TestCase):
```
Comment (ebolyen) [inline skbio/sequence/tests/test_nucleotide_sequence.py]: Remove these

File: skbio/util/_misc.py [hunk 12]
```diff
@@ -310,6 +356,34 @@ def flatten(items):
     return result
 
 
+def reprnator(start, tokens, end, separator=', '):
```
Comment (ebolyen) [inline skbio/util/_misc.py]: Need tests for reprnator

File: skbio/util/_misc.py [hunk 13]
```diff
@@ -310,6 +356,34 @@ def flatten(items):
...
+    """Live your dreams"""
```
Comment (ebolyen) [inline skbio/util/_misc.py]: Maybe a different docstring

File: skbio/alignment/_alignment.py (hunk 2 above)
Comment (gregcaporaso) [inline skbio/alignment/_alignment.py]: Oops, that was my over-zealous find/replace.

File: skbio/sequence/__init__.py [hunk 14]
```diff
@@ -39,55 +44,98 @@
...
+Sequences can be searched for known motif types. This returns the slices that
+describe the matches.
+
+>>> r5 = RNA('AGG-GGACUGAA')
+>>> for e in r5.find_motifs('purine-run', min_length=2): print(e)
+slice(0, 3, None)
+slice(4, 7, None)
+slice(9, 12, None)
+
+Those slices can be used to slice out the relevant subsequences.
```
Comment (gregcaporaso) [inline skbio/sequence/__init__.py]: *slice out* to *extract*

File: skbio/sequence/__init__.py [hunk 15]
```diff
@@ -39,55 +44,98 @@
...
+
+And gaps or other features can be ignored while searching, as these may disrupt
+otherwise meaningful motifs.
+
+>>> for e in r5.find_motifs('purine-run', min_length=2, exclude=r5.gaps()):
+...     print(r5[e])
+AGG-GGA
+GAA
+
+And removing gaps from the resulting motif matches is easily achieved, as the
```
Comment (gregcaporaso) [inline skbio/sequence/__init__.py]: *And* to *In the above example,*

File: skbio/sequence/_sequence.py [hunk 16]
```diff
@@ -9,75 +9,66 @@
...
-class BiologicalSequence(Sequence, SkbioObject):
-    """Base class for biological sequences.
+
+class Sequence(collections.Sequence, SkbioObject):
+    """Store biological sequence data and optional associated metadata.
+
+    ``Sequence`` objects do not enforce an alphabet and are thus the most
+    generic objects for storing biological sequence data. Subclasses ``DNA``,
+    ``RNA``, and ``Protein`` enforce the IUPAC character set [1]_ and provide
+    operations specific to each respective molecule type.
```
Comment (gregcaporaso) [inline skbio/sequence/_sequence.py]: *enforce the IUPAC character set [1]_ and provide operations specific to each respective molecule type.*

to 

*enforce the IUPAC character set [1]_ for, and provide operations specific to, each respective molecule type.*

File: skbio/sequence/_sequence.py [hunk 17]
```diff
@@ -86,1830 +77,1289 @@ class BiologicalSequence(Sequence, SkbioObject):
...
-        return set()
+    def _set_description(self, description):
+        if isinstance(description, str):
+            self._description = description
+        else:
+            raise TypeError("Sequence description %r must be of type `str`." %
+                            (description,))
+
+    def _set_sequence(self, sequence):
+        """Munge the sequence data into a numpy array."""
```
Comment (gregcaporaso) [inline skbio/sequence/_sequence.py]: For clarity: "Munge the sequence data into a numpy array of uint8."

File: skbio/util/_misc.py (hunk 12 above)
Comment (ebolyen) [inline skbio/util/_misc.py]: Done.

File: skbio/sequence/_sequence.py [hunk 18]
```diff
@@ -86,1830 +77,1289 @@ class BiologicalSequence(Sequence, SkbioObject):
...
+        Examples
+        --------
+        >>> from skbio import Sequence
+        >>> s = Sequence('GGUCG')
+        >>> s.count('G')
+        3
+        >>> s.count('GG')
+        1
+        >>> s.count('T')
+        0
```
Comment (johnchase) [inline skbio/sequence/_sequence.py]: An example that included the optional parameters would be nice here: 
```
s.count('G', 2, 5)
```

File: skbio/sequence/__init__.py (hunk 14 above)
Comment (jairideout) [inline skbio/sequence/__init__.py]: done

File: skbio/sequence/__init__.py (hunk 15 above)
Comment (jairideout) [inline skbio/sequence/__init__.py]: done

File: skbio/sequence/_sequence.py (hunk 16 above)
Comment (jairideout) [inline skbio/sequence/_sequence.py]: done

File: skbio/sequence/_sequence.py (hunk 17 above)
Comment (jairideout) [inline skbio/sequence/_sequence.py]: done

File: skbio/sequence/_sequence.py (hunk 18 above)
Comment (jairideout) [inline skbio/sequence/_sequence.py]: done

**Comment by ebolyen** [comment]
@gregcaporaso, @jairideout , @johnchase, and myself have reviewed this. Ready for merge when tests pass.

**Pull request merged**